src/migrate.rs

```diff
@@ -0,0 +1,964 @@
+    }
+
+    fn leaf_node_size(mn_lo: u16, mn_hi: u16) -> u32 {
+        u32::from(mn_lo) + ((u32::from(mn_hi)) << 16)
+    }
+
+    fn parse_branch_nodes(cursor: &mut Cursor<&[u8]>, pb_lower: u16, bits: Bits) -> MigrateResult<Vec<BranchNode>> {
+        let num_keys = Self::num_keys(pb_lower, bits);
+        let mp_ptrs = Self::parse_mp_ptrs(cursor, num_keys)?;
+
+        let mut branch_nodes = Vec::with_capacity(num_keys as usize);
+
+        for mp_ptr in mp_ptrs {
+            cursor.set_position(u64::from(mp_ptr));
+            branch_nodes.push(Self::parse_branch_node(cursor, bits)?)
+        }
+
+        Ok(branch_nodes)
+    }
+
+    fn parse_branch_node(cursor: &mut Cursor<&[u8]>, bits: Bits) -> MigrateResult<BranchNode> {
+        // The order of the mn_lo and mn_hi fields is endian-dependent and would
+        // be reversed in an LMDB environment created on a big-endian system.
+        let mn_lo = cursor.read_u16::<LittleEndian>()?;
+        let mn_hi = cursor.read_u16::<LittleEndian>()?;
+
+        let mn_flags = cursor.read_u16::<LittleEndian>()?;
+
+        // Branch nodes overload the mn_lo, mn_hi, and mn_flags fields
+        // to store the page number, so we derive the number from those fields.
+        let mp_pgno = Self::branch_node_page_num(mn_lo, mn_hi, mn_flags, bits);
+
+        let mn_ksize = cursor.read_u16::<LittleEndian>()?;
+
+        let position = cursor.position();
+        let start = usize::try_from(position)?;
+        let end = usize::try_from(position + u64::from(mn_ksize))?;
+        let mn_data = cursor.get_ref()[start..end].to_vec();
+        cursor.set_position(end as u64);
+
+        Ok(BranchNode {
+            mp_pgno,
+            mn_ksize,
+            mn_data,
+        })
+    }
+
+    fn branch_node_page_num(mn_lo: u16, mn_hi: u16, mn_flags: u16, bits: Bits) -> u64 {
+        let mut page_num = u64::from(u32::from(mn_lo) + (u32::from(mn_hi) << 16));
+        if bits == Bits::U64 {
+            page_num += u64::from(mn_flags) << 32;
+        }
+        page_num
+    }
+
+    fn parse_mp_ptrs(cursor: &mut Cursor<&[u8]>, num_keys: u64) -> MigrateResult<Vec<u16>> {
+        let mut mp_ptrs = Vec::with_capacity(num_keys as usize);
+        for _ in 0..num_keys {
+            mp_ptrs.push(cursor.read_u16::<LittleEndian>()?);
+        }
+        Ok(mp_ptrs)
+    }
+
+    fn num_keys(pb_lower: u16, bits: Bits) -> u64 {
+        (u64::from(pb_lower) - page_header_size(bits)) >> 1
+    }
+}
+
+pub struct Migrator {
+    file: File,
+    bits: Bits,
+}
+
+impl Migrator {
+    /// Create a new Migrator for the LMDB environment at the given path.
+    /// This tries to open the data.mdb file in the environment and determine
+    /// the bit depth of the executable that created it, so it can fail
+    /// and return an Err if the file can't be opened or the depth determined.
+    pub fn new(path: &Path) -> MigrateResult<Migrator> {
+        let mut path = PathBuf::from(path);
+        path.push("data.mdb");
```

Comment:
This path assumption won't work with the `MDB_NOSUBDIR`, as the rkv consumers are able to access this flag with `lmdb::EnvironmentFlags`. Let's add some comments to document this limitation.

Reply:
Let's also describe the supported db types in the current implementation.

Reply:
Good point, I've documented use of `Migrator`, including its limitations, in a26d01cdf82b2ebdc070b682e5bae5b456f49b20.